pipeline_dp/private_beam.py

```diff
@@ -308,3 +315,146 @@ def __init__(self, fn: Callable, label: Optional[str] = None):
+    guarantees if not implemented correctly.
+    """
+
+    @abc.abstractmethod
+    def add_input_for_private_output(self, accumulator, input):
+        """Add input, which contributes to private output.
+
+         This is a DP counterpart of `add_input()`. The same CombinerFn can
+         have both in order to be able to compute both DP and not-DP
+         aggregations.
+
+         Typically, in this function clipping is performed.
+         """
+        pass
+
+    @abc.abstractmethod
+    def extract_private_output(self, accumulator):
+        """Computes private output.
+
+        A DP mechanism with the budget requested in `request_budget()` should
+        be applied here.
+        """
+        pass
+
+    @abc.abstractmethod
+    def request_budget(self,
+                       budget_accountant: budget_accounting.BudgetAccountant):
+        """Requests the budget.
+
+        It is called by PipelineDP during the construction of the computations.
+        The custom combiner can request a DP budget by calling
+        'budget_accountant.request_budget()'. The budget object needs to be
+        stored in a field of 'self'. It will be serialized and distributed
+        to the workers together with 'self'.
+
+        Warning: do not store 'budget_accountant' in 'self'. It is assumed to
+        live in the driver process.
+        """
+        pass
+
+    def set_aggregate_params(self,
+                             aggregate_params: pipeline_dp.AggregateParams):
+        """Sets aggregate parameters
+
+        The custom combiner can optionally use it for own DP parameter
+        computations.
+        """
+        self._aggregate_params = aggregate_params
+
+
+class _CombineFnCombiner(pipeline_dp.CustomCombiner):
+
+    def __init__(self, private_combine_fn: PrivateCombineFn):
+        self._private_combine_fn = private_combine_fn
+
+    def create_accumulator(self, values):
+        """Creates accumulator from 'values'."""
+        accumulator = self._private_combine_fn.create_accumulator()
+        for v in values:
+            accumulator = self._private_combine_fn.add_private_input(
+                accumulator, v)
+        return accumulator
+
+    def merge_accumulators(self, accumulator1, accumulator2):
+        """Merges the accumulators and returns accumulator."""
+        return self._private_combine_fn.merge_accumulators(
+            [accumulator1, accumulator2])
+
+    def compute_metrics(self, accumulator):
+        """Computes and returns the result of aggregation."""
+        return self._private_combine_fn.extract_private_output(accumulator)
+
+    def request_budget(self,
+                       budget_accountant: budget_accounting.BudgetAccountant):
+        self._private_combine_fn.request_budget(budget_accountant)
+
+    def set_aggregate_params(self, aggregate_params):
+        self._private_combine_fn.set_aggregate_params(aggregate_params)
+
+
+@dataclasses.dataclass
+class CombinePerKeyParams:
```

Comment:
nit: Add docs for the class and fields, similar to other classes, e.g.: https://github.com/OpenMined/PipelineDP/blob/583aeed06dfe5cc8d49bf4fc74552cf41628cfbc/pipeline_dp/aggregate_params.py#L227